# Spam Detector

In [42]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score

## Retrieve the Data

The data is located at [https://static.bc-edx.com/ai/ail-v-1-0/m13/challenge/spam-data.csv](https://static.bc-edx.com/ai/ail-v-1-0/m13/challenge/spam-data.csv)

Dataset Source: [UCI Machine Learning Library](https://archive.ics.uci.edu/dataset/94/spambase)

Import the data using Pandas. Display the resulting DataFrame to confirm the import was successful.

In [43]:
# Import the data
data = pd.read_csv("https://static.bc-edx.com/ai/ail-v-1-0/m13/challenge/spam-data.csv")
data.head()

word_freq_make  word_freq_address  word_freq_all  word_freq_3d  \
0            0.00               0.64           0.64           0.0   
1            0.21               0.28           0.50           0.0   
2            0.06               0.00           0.71           0.0   
3            0.00               0.00           0.00           0.0   
4            0.00               0.00           0.00           0.0   

   word_freq_our  word_freq_over  word_freq_remove  word_freq_internet  \
0           0.32            0.00              0.00                0.00   
1           0.14            0.28              0.21                0.07   
2           1.23            0.19              0.19                0.12   
3           0.63            0.00              0.31                0.63   
4           0.63            0.00              0.31                0.63   

   word_freq_order  word_freq_mail  ...  char_freq_;  char_freq_(  \
0             0.00            0.00  ...         0.00        0.000   
1             0.00            0.94  ...         0.00        0.132   
2             0.64            0.25  ...         0.01        0.143   
3             0.31            0.63  ...         0.00        0.137   
4             0.31            0.63  ...         0.00        0.135   

   char_freq_[  char_freq_!  char_freq_$  char_freq_#  \
0          0.0        0.778        0.000        0.000   
1          0.0        0.372        0.180        0.048   
2          0.0        0.276        0.184        0.010   
3          0.0        0.137        0.000        0.000   
4          0.0        0.135        0.000        0.000   

   capital_run_length_average  capital_run_length_longest  \
0                       3.756                          61   
1                       5.114                         101   
2                       9.821                         485   
3                       3.537                          40   
4                       3.537                          40   

   capital_run_length_total  spam  
0                       278     1  
1                      1028     1  
2                      2259     1  
3                       191     1  
4                       191     1  

[5 rows x 58 columns]

## Predict Model Performance

You will be creating and comparing two models on this data: a Logistic Regression, and a Random Forests Classifier. Before you create, fit, and score the models, make a prediction as to which model you think will perform better. You do not need to be correct! 

Write down your prediction in the designated cells in your Jupyter Notebook, and provide justification for your educated guess.

I believe Random Forests Classifier will perform better as we have seen it outperform Logistic Regression on linear data (spam y/n) in class.  However, Random Forests can on occasion overfit data which, while it outperforms, is less reliable.

## Split the Data into Training and Testing Sets

In [44]:
# Create the labels set `y` and features DataFrame `X`
y = data['spam']
X = data.copy().drop(columns=['spam'])

In [45]:
# Check the balance of the labels variable (`y`) by using the `value_counts` function.
y.value_counts()

spam
0    2788
1    1813
Name: count, dtype: int64

In [46]:
# Split the data into X_train, X_test, y_train, y_test
X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=42)

# Check the shape of the training set
X_train.head()

word_freq_make  word_freq_address  word_freq_all  word_freq_3d  \
1173            0.00                0.0           0.00           0.0   
1954            0.67                0.0           0.00           0.0   
256             0.00                0.0           1.42           0.0   
3341            0.00                0.0           0.24           0.0   
4469            0.00                0.0           0.00           0.0   

      word_freq_our  word_freq_over  word_freq_remove  word_freq_internet  \
1173           0.00            0.00              1.23                0.00   
1954           0.00            0.00              0.00                0.00   
256            0.71            0.00              0.00                0.71   
3341           0.09            0.04              0.00                0.00   
4469           0.00            0.00              0.00                0.00   

      word_freq_order  word_freq_mail  ...  word_freq_conference  char_freq_;  \
1173             0.00            0.00  ...                   0.0        0.000   
1954             0.00            0.00  ...                   0.0        0.000   
256              0.00            0.71  ...                   0.0        0.000   
3341             0.04            0.00  ...                   0.0        0.014   
4469             0.00            0.00  ...                   0.0        0.000   

      char_freq_(  char_freq_[  char_freq_!  char_freq_$  char_freq_#  \
1173        0.000          0.0        1.939        0.000        0.000   
1954        0.055          0.0        0.000        0.000        0.000   
256         0.000          0.0        0.931        0.000        0.000   
3341        0.148          0.0        0.014        0.044        0.007   
4469        0.000          0.0        0.000        0.000        0.000   

      capital_run_length_average  capital_run_length_longest  \
1173                       1.379                           8   
1954                       1.644                          13   
256                       12.055                         117   
3341                       2.112                          26   
4469                       1.500                           2   

      capital_run_length_total  
1173                        40  
1954                        74  
256                        217  
3341                      1223  
4469                         3  

[5 rows x 57 columns]

## Scale the Features

Use the `StandardScaler` to scale the features data. Remember that only `X_train` and `X_test` DataFrames should be scaled.

In [47]:
from sklearn.preprocessing import StandardScaler

# Create the StandardScaler instance
scaler = StandardScaler()


In [48]:
# Fit the Standard Scaler with the training data
X_train_scaled = scaler.fit(X_train)


In [49]:
# Scale the training and test data
X_train_scaled = scaler.transform(X_train)
#X_test_scaled = scaler.transform(X_test)

## Create and Fit a Logistic Regression Model

Create a Logistic Regression model, fit it to the training data, make predictions with the testing data, and print the model's accuracy score. You may choose any starting settings you like. 

In [50]:
# Train a Logistic Regression model and print the model score
from sklearn.linear_model import LogisticRegression
lr_model = LogisticRegression()

# Fit the model
lr_model.fit(X_train_scaled, y_train)

# Score the model
print(f"Logistic Regression Training Score: {lr_model.score(X_train_scaled, y_train)}")
#print(f"Testing Data Score: {lr_model.score(X_test, y_test)}")

Logistic Regression Training Score: 0.9257971014492754


In [40]:
# Make and save testing predictions with the saved logistic regression model using the test data
X_test_scaled = scaler.transform(X_test)
test_predictions = lr_model.predict(X_test_scaled)

# Review the predictions
test_predictions

array([0, 0, 0, ..., 1, 0, 0])

In [41]:
# Calculate the accuracy score by evaluating `y_test` vs. `testing_predictions`.
test_accuracy = accuracy_score(y_test, test_predictions)
print(f"Logistic Regression Accuracy: {test_accuracy}")

Logistic Regression Accuracy: 0.9226759339704604


## Create and Fit a Random Forest Classifier Model

Create a Random Forest Classifier model, fit it to the training data, make predictions with the testing data, and print the model's accuracy score. You may choose any starting settings you like. 

In [28]:
# Train a Random Forest Classifier model and print the model score
from sklearn.ensemble import RandomForestClassifier

# Create a random forest classifier
rf_model = RandomForestClassifier()

#Train the model
rf_model.fit(X_train_scaled, y_train)

# Score and Print the model
print(f"Random Forest Training Score: {rf_model.score(X_train_scaled, y_train)}")

Random Forest Training Score: 0.9994202898550725


In [29]:
# Make and save testing predictions with the saved logistic regression model using the test data
rf_predictions = rf_model.predict(X_test_scaled)

# Review the predictions
rf_predictions

array([0, 0, 0, ..., 1, 0, 0])

In [30]:
# Calculate the accuracy score by evaluating `y_test` vs. `testing_predictions`.
rf_accuracy = accuracy_score(y_test, rf_predictions)

print(f"Random Forest Accuracy: {rf_accuracy}")

Random Forest Accuracy: 0.9582971329278888


## Evaluate the Models

Which model performed better? How does that compare to your prediction? Write down your results and thoughts in the following markdown cell.

As predicted the Random Forest model outperformed Logistic Regression.  Below are my results:

Logistic Regression Training Score: 0.9257971014492754
Logistic Regression Accuracy: 0.9226759339704604

Random Forest Training Score: 0.9994202898550725
Random Forest Accuracy: 0.9582971329278888
